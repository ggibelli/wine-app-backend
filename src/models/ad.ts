/* eslint-disable @typescript-eslint/no-unsafe-member-access */
import mongoose, { Schema, Document } from 'mongoose';
// import { Address } from '../types';
import {
  // Province,
  // Regioni,
  METODOPRODUTTIVO,
} from '../utils/enumMongooseHelper';
import { IUserDoc } from './user';
import { IWineDoc } from './wine';
import { TypeAd, TypeProduct, Menzione } from '../types';
import { IVineyardDoc } from './vineyard';
import { MetodoProduttivo } from '../types';
import { Negotiation } from '../generated/graphql';

export interface IAd {
  postedBy: IUserDoc['_id'];
  wineName?: string;
  wine?: IWineDoc['_id'];
  sottoZona?: string;
  menzione?: Menzione;
  metodoProduttivo?: MetodoProduttivo;
  vineyardName?: string;
  vineyard?: IVineyardDoc['_id'];
  harvest: number;
  abv: number;
  priceFrom: number; //se vendita priceFrom e priceTo settati a stesso numero
  priceTo: number;
  litersFrom?: number; //se vendite litersFrom e litersTo settati a stesso numero
  litersTo?: number;
  kgFrom?: number;
  kgTo?: number;
  content?: string | null | undefined;
  // address: Address;
  negotiations?: mongoose.Types.Array<Negotiation>; // trattative dell'annuncio, solo attive, solo graphql??
  viewedBy?: mongoose.Types.Array<IUserDoc['_id']>;
  savedBy?: mongoose.Types.Array<IUserDoc['_id']>;
  typeAd: TypeAd;
  typeProduct: TypeProduct;
  needsFollowUp: boolean;
  isActive: boolean;
  datePosted: Date;
}

export interface AdGraphQl extends IAd {
  _id: mongoose.Types.ObjectId | string;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const adSchemaFields: Record<keyof IAd, any> = {
  postedBy: {
    type: Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  typeAd: {
    type: String,
    enum: [TypeAd.SELL, TypeAd.BUY],
    required: true,
  },
  typeProduct: {
    type: String,
    enum: [TypeProduct.ADWINE, TypeProduct.ADGRAPE],
    required: true,
  },
  wineName: {
    type: String,
    minlength: 3,
    required() {
      return this.typeProduct === TypeProduct.ADWINE;
    },
    index: true,
  },
  wine: {
    type: Schema.Types.ObjectId,
    ref: 'Wine',
    // required() {
    //   return this.typeProduct === TypeProduct.ADWINE;
    // },
  },
  sottoZona: {
    type: String,
  },
  menzione: {
    type: String,
    enum: ['CLASSICO', 'RISERVA', 'SUPERIORE', 'VIGNA'],
  },
  metodoProduttivo: {
    type: String,
    enum: Object.values(METODOPRODUTTIVO),
    required: true,
    default: 'CONVENZIONALE',
  },
  vineyardName: {
    type: String,
    minlength: 3,
    required() {
      return this.typeProduct === TypeProduct.ADGRAPE;
    },
    index: true,
  },
  vineyard: {
    type: Schema.Types.ObjectId,
    ref: 'Vineyard',
    // required() {
    //   return this.typeProduct === TypeProduct.ADGRAPE;
    // },
  },
  abv: Number,
  harvest: {
    type: Number,
    required: true,
  },
  priceFrom: {
    type: Number,
    required: true,
  },
  priceTo: {
    type: Number,
    required: true,
  },
  litersFrom: {
    type: Number,
    required() {
      return this.typeProduct === TypeProduct.ADWINE;
    },
  },
  litersTo: {
    type: Number,
    required() {
      return this.typeProduct === TypeProduct.ADWINE;
    },
  },
  kgFrom: {
    type: Number,
    required() {
      return this.typeProduct === TypeProduct.ADGRAPE;
    },
  },
  kgTo: {
    type: Number,
    required() {
      return this.typeProduct === TypeProduct.ADGRAPE;
    },
  },
  // address: {
  //   comune: {
  //     type: String,
  //     required: true,
  //   },
  //   provincia: {
  //     type: String,
  //     enum: Object.values(Province),
  //     required: true,
  //   },
  //   regione: {
  //     type: String,
  //     enum: Object.values(Regioni),
  //     required: true,
  //   },
  // },
  negotiations: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Negotiation',
    },
  ],
  viewedBy: [
    {
      type: Schema.Types.ObjectId,
      ref: 'User',
    },
  ],
  savedBy: [
    {
      type: Schema.Types.ObjectId,
      ref: 'User',
    },
  ],
  content: {
    type: String,
  },
  isActive: {
    type: Boolean,
    default: true,
  },
  needsFollowUp: {
    type: Boolean,
    default: true,
  },
  datePosted: {
    type: Date,
    // `Date.now()` returns the current unix timestamp as a number
    default: Date.now,
  },
};

export interface IAdDoc extends IAd, Document {}

const adSchema = new Schema(adSchemaFields);

adSchema.index({ typeProduct: 1, typeAd: 1, wineName: 1, vineyardName: 1 });

export const Ad = mongoose.model<IAdDoc>('Ad', adSchema);
