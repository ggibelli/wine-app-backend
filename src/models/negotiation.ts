import mongoose, { Schema, Document } from 'mongoose';
import { IUserDoc } from './user';
import { IAdDoc } from './ad';
import { IMessageDoc } from './message';
import { TypeAd } from '../types';
import mongooseUniqueValidator from 'mongoose-unique-validator';

export interface INegotiation {
  createdBy: IUserDoc['_id'];
  ad: IAdDoc['_id'];
  forUserAd: IUserDoc['_id'];
  messages?: mongoose.Types.Array<IMessageDoc['_id']>;
  isConcluded: boolean;
  dateCreated: Date;
  type: TypeAd;
}

export interface NegotiationGraphQl extends INegotiation {
  _id: mongoose.Types.ObjectId;
}

export interface INegotiationDoc extends INegotiation, Document {}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const negotiationSchemaFields: Record<keyof INegotiation, any> = {
  createdBy: {
    type: Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  type: {
    type: String,
    enum: ['SELL', 'BUY'],
    required: true,
  },
  ad: {
    type: Schema.Types.ObjectId,
    ref: 'Ad',
    required: true,
  },
  forUserAd: {
    type: Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  messages: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Message',
    },
  ],
  isConcluded: {
    type: Boolean,
    default: false,
  },
  dateCreated: {
    type: Date,
    // `Date.now()` returns the current unix timestamp as a number
    default: Date.now,
  },
};

const negotiationSchema = new Schema(negotiationSchemaFields);

negotiationSchema.index({ createdBy: 1, ad: 1 }, { unique: true });
negotiationSchema.plugin(mongooseUniqueValidator);
export const Negotiation = mongoose.model<INegotiationDoc>(
  'Negotiation',
  negotiationSchema
);
