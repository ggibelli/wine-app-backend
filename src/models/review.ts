import mongoose, { Schema, Document } from 'mongoose';
import { IUserDoc } from './user';
import { INegotiationDoc } from './negotiation';
import { TypeAd, Rating } from '../types';
import mongooseUniqueValidator from 'mongoose-unique-validator';

export interface IReview {
  createdBy: IUserDoc['_id'];
  negotiation: INegotiationDoc['_id'];
  forUser: IUserDoc['_id'];
  rating: Rating;
  dateCreated: Date;
  content: string;
  type: TypeAd;
}

export interface ReviewGraphQl extends IReview {
  _id: mongoose.Types.ObjectId;
}

export interface IReviewDoc extends IReview, Document {}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const reviewSchemaFields: Record<keyof IReview, any> = {
  createdBy: {
    type: Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  negotiation: {
    type: Schema.Types.ObjectId,
    ref: 'Negotiation',
    required: true,
  },
  forUser: {
    type: Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  dateCreated: {
    type: Date,
    // `Date.now()` returns the current unix timestamp as a number
    default: Date.now,
  },
  content: {
    type: String,
    minlength: 5,
  },
  rating: {
    type: String,
    enum: ['POOR', 'AVERAGE', 'OK', 'GOOD', 'PERFECT'],
    required: true,
  },
  type: {
    type: String,
    enum: ['SELL', 'BUY'],
  },
};

const reviewSchema = new Schema(reviewSchemaFields);

reviewSchema.index({ createdBy: 1, negotiation: 1 }, { unique: true });
reviewSchema.plugin(mongooseUniqueValidator);

export const Review = mongoose.model<IReviewDoc>('Review', reviewSchema);
