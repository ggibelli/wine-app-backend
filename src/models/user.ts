/* eslint-disable @typescript-eslint/no-unsafe-return */
import mongoose, { Schema, Document, HookNextFunction } from 'mongoose';
import mongooseUniqueValidator from 'mongoose-unique-validator';
import { Address } from '../types';
import { Province, Regioni } from '../utils/enumMongooseHelper';
import { IAdDoc } from './ad';
import { IWineDoc } from './wine';
import { IReviewDoc } from './review';
import { IVineyardDoc } from './vineyard';
import { MetodoProduttivo } from '../types';
import { METODOPRODUTTIVO } from '../utils/enumMongooseHelper';
import bcrypt from 'bcrypt';
import { IMessageDoc } from './message';
import { Negotiation } from '../generated/graphql';

const HASH_ROUNDS = 10;

interface ProducedWines {
  wine: IWineDoc['_id'] | IWineDoc;
  bottlesProduced: number;
  metodoProduttivo: MetodoProduttivo;
}

interface OwnedVineyards {
  vineyard: IVineyardDoc['_id'] | IVineyardDoc;
  tonsProduced: number;
  metodoProduttivo: MetodoProduttivo;
}

export interface Coordinates {
  latitude: number;
  longitude: number;
}

export interface IUser {
  email: string;
  firstName: string;
  lastName: string;
  password: string;
  pIva: string;
  phoneNumber: string;
  address: Address;
  coordinates?: Coordinates;
  isVerified: boolean;
  isPremium?: boolean;
  isAdmin: boolean;
  hideContact: boolean;
  ads?: mongoose.Types.Array<IAdDoc['_id'] | IAdDoc>; // annunci postati dall'utente
  savedAds?: mongoose.Types.Array<IAdDoc['_id'] | IAdDoc>; // annunci postati dall'utente
  messages?: mongoose.Types.Array<IMessageDoc['_id'] | IMessageDoc>;
  negotiations?: mongoose.Types.Array<Negotiation>; // trattative dell'utente, attive e non, concluse e non
  reviews?: mongoose.Types.Array<IReviewDoc['_id'] | IReviewDoc>; // recensioni fatte e ricevute dall'utente
  adsRemaining?: number;
  dateCreated: Date;
  producedWines?: mongoose.Types.Array<ProducedWines>;
  ownedVineyards?: mongoose.Types.Array<OwnedVineyards>;
}

export interface IUserDoc extends IUser, Document {
  validatePassword(password: string): Promise<boolean>;
}

export interface UserGraphQl extends IUser {
  validatePassword(password: string): Promise<boolean>;
  _id: mongoose.Types.ObjectId;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const userSchemaFields: Record<keyof IUser, any> = {
  email: {
    type: String,
    unique: true,
    minlength: 3,
    required: true,
  },
  phoneNumber: {
    type: String,
    unique: true,
    minlength: 10,
    required: true,
  },
  firstName: String,
  lastName: String,
  password: String,
  pIva: {
    type: String,
    unique: true,
    minlength: 11,
    required: true,
  },
  address: {
    via: {
      type: String,
      required: true,
      minlength: 5,
    },
    CAP: {
      type: String,
      required: true,
      minlength: 5,
    },
    comune: { type: String, required: true, minlength: 5 },
    provincia: {
      type: String,
      enum: Object.values(Province),
      required: true,
    },
    regione: {
      type: String,
      enum: Object.values(Regioni),
      required: true,
    },
  },
  coordinates: {
    latitude: Number,
    longitude: Number,
  },
  isVerified: {
    type: Boolean,
    required: true,
    default: false,
  },
  isPremium: {
    type: Boolean,
    default: false,
  },
  isAdmin: {
    type: Boolean,
    default: false,
  },
  hideContact: {
    type: Boolean,
    default: true,
  },
  ads: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Ad',
    },
  ],
  savedAds: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Ad',
    },
  ],
  messages: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Message',
    },
  ],
  negotiations: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Negotiation',
    },
  ],
  reviews: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Review',
    },
  ],
  adsRemaining: Number,
  dateCreated: {
    type: Date,
    // `Date.now()` returns the current unix timestamp as a number
    default: Date.now,
  },
  producedWines: [
    {
      wine: {
        type: Schema.Types.ObjectId,
        ref: 'Wine',
      },
      bottlesProduced: Number,
      metodoProduttivo: {
        type: String,
        enum: Object.values(METODOPRODUTTIVO),
      },
    },
  ],
  ownedVineyards: [
    {
      vineyard: {
        type: Schema.Types.ObjectId,
        ref: 'Vineyard',
      },
      tonsProduced: Number,
      metodoProduttivo: {
        type: String,
        enum: Object.values(METODOPRODUTTIVO),
      },
    },
  ],
};

const userSchema = new Schema(userSchemaFields);

userSchema.pre('save', async function (next: HookNextFunction) {
  // here we need to retype 'this' because by default it is
  // of type Document from which the 'IUser' interface is inheriting
  // but the Document does not know about our password property
  const thisObj = this as IUserDoc;

  if (!this.isModified('password')) {
    return next();
  }

  try {
    const salt = await bcrypt.genSalt(HASH_ROUNDS);
    thisObj.password = await bcrypt.hash(thisObj.password, salt);
    return next();
  } catch (e) {
    return next(e);
  }
});

userSchema.methods.validatePassword = async function (pass: string) {
  return bcrypt.compare(pass, this.password);
};

userSchema.plugin(mongooseUniqueValidator);

export const User = mongoose.model<IUserDoc>('User', userSchema);
