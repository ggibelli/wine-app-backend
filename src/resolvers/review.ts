/* eslint-disable @typescript-eslint/no-unsafe-call */
import {
  Resolvers,
  ReviewInput,
  ReviewInputUpdate,
} from '../generated/graphql';
import Users from '../data-sources/users';
import Reviews from '../data-sources/reviews';
import Ads from '../data-sources/ads';
import Messages from '../data-sources/messages';
import Negotiations from '../data-sources/negotiations';
import { PubSub, withFilter } from 'apollo-server-express';
import { ReviewGraphQl } from '../models/review';
import { UserGraphQl } from '../models/user';

const pubsub = new PubSub();

const REVIEW_CREATED = 'REVIEW_CREATED';

interface StringIndexSignatureInterface {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  [index: string]: any;
}

type StringIndexed<T> = T & StringIndexSignatureInterface;

interface MongoDataSource {
  users: Users;
  reviews: Reviews;
  ads: Ads;
  messages: Messages;
  negotiations: Negotiations;
}

export const resolver: StringIndexed<Resolvers> = {
  Query: {
    async reviews(_, __, { dataSources }: { dataSources: MongoDataSource }) {
      return dataSources.reviews.getReviews();
    },
    async review(_, { id }, { dataSources }: { dataSources: MongoDataSource }) {
      return dataSources.reviews.getReview(id);
    },
  },
  Mutation: {
    async createReview(
      _,
      { review }: { review: ReviewInput },
      { dataSources }: { dataSources: MongoDataSource }
    ) {
      const reviewResponse = await dataSources.reviews.createReview(review);
      if (!reviewResponse.errors.length) {
        await pubsub.publish(REVIEW_CREATED, {
          reviewCreated: reviewResponse.response,
        });
      }
      return reviewResponse;
    },

    async updateReview(
      _,
      { review }: { review: ReviewInputUpdate },
      { dataSources }: { dataSources: MongoDataSource }
    ) {
      return dataSources.reviews.updateReview(review);
    },
    async deleteReview(
      _,
      { id }: { id: string },
      { dataSources }: { dataSources: MongoDataSource }
    ) {
      return dataSources.reviews.deleteReview(id);
    },
  },

  Subscription: {
    reviewCreated: {
      subscribe: withFilter(
        () => pubsub.asyncIterator([REVIEW_CREATED]),
        (
          payload: { reviewCreated: ReviewGraphQl },
          _,
          { user }: { user: UserGraphQl }
        ) => {
          return Boolean(
            // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
            payload.reviewCreated.forUser._id.toString() ===
              user._id.toHexString()
          );
        }
      ),
    },
  },

  Review: {
    async createdBy(
      review,
      _,
      { dataSources }: { dataSources: MongoDataSource }
    ) {
      return dataSources.users.getUser(review.createdBy);
    },
    async negotiation(
      review,
      _,
      { dataSources }: { dataSources: MongoDataSource }
    ) {
      return dataSources.negotiations.getNegotiation(review.negotiation);
    },
    async forUser(
      review,
      _,
      { dataSources }: { dataSources: MongoDataSource }
    ) {
      return dataSources.users.getUser(review.forUser);
    },
  },
};
