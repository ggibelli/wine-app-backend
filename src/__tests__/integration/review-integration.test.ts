/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
import { gql } from 'apollo-server-express';
import cron from 'cron';
import { LeanDocument, Types } from 'mongoose';
import {
  testClient,
  connectToDb,
  dropTestDb,
  closeDbConnection,
} from '../../tests/integrationSetup';
// import { ObjectId } from 'mongodb';
import { User } from '../../models/user';
import { Ad } from '../../models/ad';
import { Review } from '../../models/review';
import { users, ads } from '../../tests/mocksTests';
import { ReviewInput, TypeAd } from '../../generated/graphql';
import { Negotiation } from '../../models/negotiation';
import {
  ReviewDocument,
  NegotiationDocument,
} from '../../interfaces/mongoose.gen';

const fakeStop = jest.fn(() => null);
const fakeStart = jest.fn(() => fakeStop());
const FakeCron = jest.fn(() => ({
  stop: fakeStop,
  start: fakeStart,
  fireOnTick: fakeStart,
}));
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
cron.CronJob = FakeCron;

// import { ObjectId } from 'mongodb';

const { query, mutate, setOptions } = testClient;

const REVIEWS = gql`
  {
    reviews {
      reviews {
        createdBy {
          firstName
        }
        negotiation {
          type
        }
        forUser {
          firstName
        }
        rating
      }
    }
  }
`;

const REVIEW = gql`
  query GetReview($id: ID!) {
    review(id: $id) {
      createdBy {
        firstName
      }
      negotiation {
        type
      }
      forUser {
        firstName
      }
      rating
    }
  }
`;

const CREATE_REVIEW = gql`
  mutation CreateReview($review: ReviewInput!) {
    createReview(review: $review) {
      response {
        createdBy {
          firstName
        }
        negotiation {
          type
        }
        forUser {
          firstName
        }
        rating
      }
      errors {
        name
        text
      }
    }
  }
`;

const UPDATE_REVIEW = gql`
  mutation UpdateReview($review: ReviewInputUpdate!) {
    updateReview(review: $review) {
      response {
        createdBy {
          firstName
        }
        negotiation {
          type
        }
        forUser {
          firstName
        }
        rating
      }
      errors {
        name
        text
      }
    }
  }
`;

const DELETE_REVIEW = gql`
  mutation DeleteReview($id: ID!) {
    deleteReview(id: $id) {
      response {
        createdBy {
          firstName
        }
        negotiation {
          type
        }
        forUser {
          firstName
        }
        rating
      }
      errors {
        name
        text
      }
    }
  }
`;

const LOGIN_VALID = gql`
  mutation {
    login(email: "gio@prova.it", password: "giovanni") {
      response {
        token
      }
      errors {
        name
        text
      }
    }
  }
`;

const LOGIN_VALID_OTHER = gql`
  mutation {
    login(email: "gio2@prova.it", password: "giovanni") {
      response {
        token
      }
      errors {
        name
        text
      }
    }
  }
`;

beforeAll(async () => {
  await connectToDb();
  await dropTestDb();
  const usersMock = users();
  const adsMock = ads();
  const user = new User({ ...usersMock[0], _id: new Types.ObjectId() });
  const otherUser = new User({ ...usersMock[1], _id: new Types.ObjectId() });
  const thirdUser = new User({ ...usersMock[2], _id: new Types.ObjectId() });
  const ad = new Ad({
    ...adsMock[0],
    postedBy: user,
    _id: new Types.ObjectId(),
  });
  const otherAd = new Ad({
    ...adsMock[1],
    postedBy: otherUser,
    _id: new Types.ObjectId(),
  });
  // const thirdAd = new Ad({ ...adsMock[2], postedBy: thirdUser });
  const negotiation = new Negotiation({
    createdBy: otherUser,
    ad,
    forUserAd: ad.postedBy,
    type: ad.typeAd,
    _id: new Types.ObjectId(),
  });
  const otherNegotiation = new Negotiation({
    createdBy: thirdUser,
    ad: otherAd,
    forUserAd: otherAd.postedBy,
    type: otherAd.typeAd,
    _id: new Types.ObjectId(),
  });
  const review = new Review({
    createdBy: otherUser,
    negotiation,
    forUser: negotiation.forUserAd,
    type: ad.typeAd,
    rating: 5,
    content: 'perfect',
    _id: new Types.ObjectId(),
  });
  const otherReview = new Review({
    createdBy: thirdUser,
    negotiation: otherNegotiation,
    forUser: otherNegotiation.forUserAd,
    type: ad.typeAd,
    rating: 2,
    content: 'very poor',
    _id: new Types.ObjectId(),
  });
  await user.save();
  await otherUser.save();
  await thirdUser.save();
  await ad.save();
  await otherAd.save();
  await negotiation.save();
  await otherNegotiation.save();
  await review.save();
  await otherReview.save();
});

describe('Integration test reviews', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });
  it('query reviews fails if not logged in', async () => {
    const res = await query(REVIEWS);
    expect(res).toMatchSnapshot();
  });

  it('query single review fails if not logged', async () => {
    const review: LeanDocument<ReviewDocument>[] = await Review.find({})
      .lean()
      .exec();
    const res = await query(REVIEW, {
      variables: { id: review[0]._id.toString() },
    });
    expect(res).toMatchSnapshot();
  });

  it('create review mutation fails if not logged in', async () => {
    const negotiation: LeanDocument<NegotiationDocument>[] =
      await Negotiation.find({}).lean().exec();
    const review: ReviewInput = {
      negotiation: negotiation[0]._id.toString(),
      forUser: negotiation[0].forUserAd.toString(),
      type: negotiation[0].type as TypeAd,
      rating: 4,
      content: 'super good',
    };
    const res = await mutate(CREATE_REVIEW, {
      variables: { review },
    });
    expect(res).toMatchSnapshot();
  });

  it('create review mutation successfull', async () => {
    const data: any = await mutate(LOGIN_VALID);
    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
    const { token } = data.data.login.response;
    setOptions({
      request: {
        headers: {
          authorization: token,
        },
      },
    });
    const negotiation: LeanDocument<NegotiationDocument>[] =
      await Negotiation.find({}).lean().exec();
    const review: ReviewInput = {
      negotiation: negotiation[1]._id.toString(),
      forUser: negotiation[1].forUserAd.toString(),
      type: negotiation[1].type as TypeAd,
      rating: 3,
      content: 'average',
    };
    const res = await mutate(CREATE_REVIEW, {
      variables: { review },
    });
    expect(res).toMatchSnapshot();
    expect(fakeStart).toBeCalledTimes(1);
    expect(fakeStop).toBeCalledTimes(1);
  });

  it('create review mutation fails if same negotiation and user', async () => {
    const negotiation: LeanDocument<NegotiationDocument>[] =
      await Negotiation.find({}).lean().exec();
    const review: ReviewInput = {
      negotiation: negotiation[1]._id.toString(),
      forUser: negotiation[1].forUserAd.toString(),
      type: negotiation[1].type as TypeAd,
      rating: 3,
      content: 'average',
    };
    const res: any = await mutate(CREATE_REVIEW, {
      variables: { review },
    });
    expect(res.data.createReview.response).toBeNull();
    expect(res.data.createReview.errors).toHaveLength(1);
  });

  it('create review mutation fails if is for own user', async () => {
    const negotiation: LeanDocument<NegotiationDocument>[] =
      await Negotiation.find({}).lean().exec();
    const review: ReviewInput = {
      negotiation: negotiation[0]._id.toString(),
      forUser: negotiation[0].forUserAd.toString(),
      type: negotiation[0].type as TypeAd,
      rating: 2,
      content: 'average',
    };
    const res = await mutate(CREATE_REVIEW, {
      variables: { review },
    });
    expect(res).toMatchSnapshot();
  });

  it('query reviews succeds if logged in and admin, all are shown', async () => {
    const res = await query(REVIEWS);
    expect(res).toMatchSnapshot();
  });

  it('query single review succeds if logged', async () => {
    const review: LeanDocument<ReviewDocument>[] = await Review.find({})
      .lean()
      .exec();
    const res = await query(REVIEW, {
      variables: { id: review[0]._id.toString() },
    });
    expect(res).toMatchSnapshot();
  });

  it('update review mutation succeds if logged in and same user', async () => {
    const user = await User.findOne({ firstName: 'Giovanni' });
    if (!user) throw new Error();
    const reviewToUpdate: LeanDocument<ReviewDocument> | null =
      await Review.findOne({
        createdBy: user,
      })
        .lean()
        .exec();
    const review = {
      _id: reviewToUpdate?._id.toString(),
      content: 'cambiato',
    };
    const res = await mutate(UPDATE_REVIEW, {
      variables: { review },
    });
    expect(res).toMatchSnapshot();
  });

  it('update review mutation fails if logged in and not same user', async () => {
    const user = await User.findOne({ firstName: 'Mariuccio' });
    if (!user) throw new Error();

    const reviewToUpdate: LeanDocument<ReviewDocument> | null =
      await Review.findOne({
        createdBy: user,
      })
        .lean()
        .exec();
    const review = {
      _id: reviewToUpdate?._id.toString(),
      content: 'non cambiato',
    };
    const res = await mutate(UPDATE_REVIEW, {
      variables: { review },
    });
    expect(res).toMatchSnapshot();
  });

  it('delete review mutation succeds if logged in and same user', async () => {
    const user = await User.findOne({ firstName: 'Giovanni' });
    if (!user) throw new Error();

    const reviewToDelete: LeanDocument<ReviewDocument> | null =
      await Review.findOne({
        createdBy: user,
      })
        .lean()
        .exec();
    const res = await mutate(DELETE_REVIEW, {
      variables: { id: reviewToDelete?._id.toString() },
    });
    expect(res).toMatchSnapshot();
  });

  it('delete review mutation fails if logged in and not same user', async () => {
    const user = await User.findOne({ firstName: 'Mariuccio' });
    if (!user) throw new Error();

    const reviewToDelete: LeanDocument<ReviewDocument> | null =
      await Review.findOne({
        createdBy: user,
      })
        .lean()
        .exec();
    const res = await mutate(DELETE_REVIEW, {
      variables: { id: reviewToDelete?._id.toString() },
    });
    expect(res).toMatchSnapshot();
  });

  it('query reviews succeds if logged in and not admin, only user review shown', async () => {
    const data: any = await mutate(LOGIN_VALID_OTHER);
    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
    const { token } = data.data.login.response;
    setOptions({
      request: {
        headers: {
          authorization: token,
        },
      },
    });
    const res = await query(REVIEWS);
    expect(res).toMatchSnapshot();
  });
});

afterAll(async () => {
  await dropTestDb();
  await closeDbConnection();
});
