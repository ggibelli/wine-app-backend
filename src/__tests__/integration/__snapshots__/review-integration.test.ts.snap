// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration test reviews create review mutation fails if is for own user 1`] = `
Object {
  "data": Object {
    "createReview": Object {
      "errors": Array [
        Object {
          "name": "General Error",
          "text": "You cannot review yourself",
        },
      ],
      "response": null,
    },
  },
}
`;

exports[`Integration test reviews create review mutation fails if not logged in 1`] = `
Object {
  "data": Object {
    "createReview": null,
  },
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "UNAUTHENTICATED",
      },
      "locations": Array [
        Object {
          "column": 3,
          "line": 2,
        },
      ],
      "message": "You need to login to continue",
      "path": Array [
        "createReview",
      ],
    },
  ],
}
`;

exports[`Integration test reviews create review mutation successfull 1`] = `
Object {
  "data": Object {
    "createReview": Object {
      "errors": Array [],
      "response": Object {
        "createdBy": Object {
          "firstName": "Giovanni",
        },
        "forUser": Object {
          "firstName": "Mariuccio",
        },
        "negotiation": Object {
          "type": "SELL",
        },
        "rating": "AVERAGE",
      },
    },
  },
}
`;

exports[`Integration test reviews delete review mutation fails if logged in and not same user 1`] = `
Object {
  "data": Object {
    "deleteReview": Object {
      "errors": Array [
        Object {
          "name": "General error",
          "text": "Errors during the review delete, only creator can delete review",
        },
      ],
      "response": null,
    },
  },
}
`;

exports[`Integration test reviews delete review mutation succeds if logged in and same user 1`] = `
Object {
  "data": Object {
    "deleteReview": Object {
      "errors": Array [],
      "response": Object {
        "createdBy": Object {
          "firstName": "Giovanni",
        },
        "forUser": Object {
          "firstName": "Mariuccio",
        },
        "negotiation": Object {
          "type": "SELL",
        },
        "rating": "AVERAGE",
      },
    },
  },
}
`;

exports[`Integration test reviews query reviews fails if not logged in 1`] = `
Object {
  "data": Object {
    "reviews": null,
  },
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "UNAUTHENTICATED",
      },
      "locations": Array [
        Object {
          "column": 3,
          "line": 2,
        },
      ],
      "message": "You need to login to continue",
      "path": Array [
        "reviews",
      ],
    },
  ],
}
`;

exports[`Integration test reviews query reviews succeds if logged in and admin, all are shown 1`] = `
Object {
  "data": Object {
    "reviews": Array [
      Object {
        "createdBy": Object {
          "firstName": "Mariuccio",
        },
        "forUser": Object {
          "firstName": "Giovanni",
        },
        "negotiation": Object {
          "type": "BUY",
        },
        "rating": "PERFECT",
      },
      Object {
        "createdBy": Object {
          "firstName": "Luigetto",
        },
        "forUser": Object {
          "firstName": "Mariuccio",
        },
        "negotiation": Object {
          "type": "SELL",
        },
        "rating": "POOR",
      },
      Object {
        "createdBy": Object {
          "firstName": "Giovanni",
        },
        "forUser": Object {
          "firstName": "Mariuccio",
        },
        "negotiation": Object {
          "type": "SELL",
        },
        "rating": "AVERAGE",
      },
    ],
  },
}
`;

exports[`Integration test reviews query reviews succeds if logged in and not admin, only user review shown 1`] = `
Object {
  "data": Object {
    "reviews": Array [
      Object {
        "createdBy": Object {
          "firstName": "Mariuccio",
        },
        "forUser": Object {
          "firstName": "Giovanni",
        },
        "negotiation": Object {
          "type": "BUY",
        },
        "rating": "PERFECT",
      },
      Object {
        "createdBy": Object {
          "firstName": "Luigetto",
        },
        "forUser": Object {
          "firstName": "Mariuccio",
        },
        "negotiation": Object {
          "type": "SELL",
        },
        "rating": "POOR",
      },
    ],
  },
}
`;

exports[`Integration test reviews query single review fails if not logged 1`] = `
Object {
  "data": Object {
    "review": null,
  },
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "UNAUTHENTICATED",
      },
      "locations": Array [
        Object {
          "column": 3,
          "line": 2,
        },
      ],
      "message": "You need to login to continue",
      "path": Array [
        "review",
      ],
    },
  ],
}
`;

exports[`Integration test reviews query single review succeds if logged 1`] = `
Object {
  "data": Object {
    "review": Object {
      "createdBy": Object {
        "firstName": "Mariuccio",
      },
      "forUser": Object {
        "firstName": "Giovanni",
      },
      "negotiation": Object {
        "type": "BUY",
      },
      "rating": "PERFECT",
    },
  },
}
`;

exports[`Integration test reviews update review mutation fails if logged in and not same user 1`] = `
Object {
  "data": Object {
    "updateReview": Object {
      "errors": Array [
        Object {
          "name": "General error",
          "text": "Errors during the review update, only creator can update review",
        },
      ],
      "response": null,
    },
  },
}
`;

exports[`Integration test reviews update review mutation succeds if logged in and same user 1`] = `
Object {
  "data": Object {
    "updateReview": Object {
      "errors": Array [],
      "response": Object {
        "createdBy": Object {
          "firstName": "Giovanni",
        },
        "forUser": Object {
          "firstName": "Mariuccio",
        },
        "negotiation": Object {
          "type": "SELL",
        },
        "rating": "AVERAGE",
      },
    },
  },
}
`;
